[v8.11] [CI/gitlab] enable all windows addons and 32-bit builds

vbgl opened the pull request:
See the corresponding commit on branch “v8.10”: https://github.com/coq/coq/commit/0efcfffd31b3d169527f5cd249676e4319a598e7

This ensures that at release time, we are sure that we know how to build the installers for Windows.

ppedrot commented:
This seems broken...

vbgl commented:
The build failure reveals a bug (see: https://coq.gitlab.io/-/coq/-/jobs/365043605/artifacts/artifacts/buildlogs/vst-a04b451b3ef9fd99007115f7745713f6fc84d1dc-patch_log.txt):

> patching file Makefile
Hunk #1 FAILED at 82.
1 out of 1 hunk FAILED -- saving rejects to file Makefile.rej

Zimmi48 commented:
@MSoegtropIMC might be able to help debug the Windows-only VST failure. If that's too much to do for the beta, then VST should be commented out from the list of add-on, and this PR should still get merged.

ppedrot commented:
This needs to be done quickly, the release is supposed to occur today-ish. What's the file to modify?

Zimmi48 commented:
Here:

https://github.com/coq/coq/blob/639886e771882e2c3302cbadc35e9383b34db1fe/dev/ci/gitlab.bat#L40-L60

Zimmi48 commented:
Interval is broken too (not surprising because not tested in CI) and should thus be removed as well. Will push a commit doing this: https://coq.gitlab.io/-/coq/-/jobs/366040015/artifacts/artifacts/buildlogs/interval-839a03e1bddbafab868fbceee59abe678e32a0f3-remake_err.txt

Zimmi48 commented:
Done. I only commented out these add-ons rather than removing them entirely from the list, because they should be added again once fixed.

ppedrot commented:
Seems very broken now...

ppedrot commented:
It seems that you can't have single-line comments that are not treated as one-line no-op commands in bat files :suspect:. You can maybe add the comment on the line after this block...

Zimmi48 commented:
I wasn't aware of this. Thanks @vbgl for fixing the mess.

MSoegtropIMC commented:
Hi, I will have a look at this. It is in general not trivial to find a pick of the add ons delivered with the Windows installer which does actually work - it never was. I will look into it asap.

And of cause it is time to continue the work on the Coq platform to decouple Coq releases from all the plugins.

I will give my assessment on how much work this will likely be early afternoon.

MSoegtropIMC commented:
A short note on Interval: this is an essential for anybody doing numerics or robotics with Coq. It generates proofs in s second which would take days to do manually. Unless the issues are severe, I would prefer to keep it.

Zimmi48 commented:
Is it really a big deal if it is not in the beta as long as it is in the final release?

MSoegtropIMC commented:
> Is it really a big deal if it is not in the beta as long as it is in the final release?

Well literally none of my developments would work without it, so I couldn't use the beta for daily work which would substantially reduce the strength of my tests. I would say it depends on how large the beta tester community for Windows is if this is an issue or not.

ppedrot commented:
@MSoegtropIMC I'll delay the release until your assessment, and we'll improvise then.

MSoegtropIMC commented:
I did run a local build and had a quick look:

- the issue with VST should be easy to fix, this is just a hack in the makefile which I need to make the installations of CompCert and VST coexist (so that VST picks up its own copy of CompCert files and not the main CompCert files).

- the issues in Interval and GappaPlugin I need to look into - from the error messages it is not easy to see if it would be trivial to fix or not. But I know both developments to some depth and will try to provide a patch until tomorrow. I guess for the beta a patch is OK. I will work with the authors to get them upstream in case I can fix it.

MSoegtropIMC commented:
P.S.: it is a bug that the VST build runs through despite the patching error - I thought I fixed this a while back - need to look into this again.

ppedrot commented:
OK, will wait for your input then.

vbgl commented:
> P.S.: it is a bug that the VST build runs through despite the patching error - I thought I fixed this a while back - need to look into this again.

The patch for VST I have fixed: this explains that the build succeed.

MSoegtropIMC commented:
The issues in Interval seem to come from the fact that in 8.11 there are two scopes delimited by **N**, the N_scope of the standard libary (which afaik used to be delimited with num) and the nat_scope of ssreflect. E.g. in 8.11 a Print Scopes contains:

```
Scope N_scope
Delimiting key is N
Bound to classes N N.t
"x 'mod' y" := N.modulo x y
"x < y" := N.lt x y

Scope nat_scope
Delimiting key is N
Bound to classes nat Nat.t
"x 'mod' y" := Nat.modulo x y
"m < n" := leq (S m) n
```

In 8.10.2 these scoped have different delimiters. So in 8.11. here and there Interval picks up the wrong scope and the proof script ends up to be for the wrong proof goal or vice versa.

I guess I can find ways to fix this wrangling the scope contexts, but I wonder if this is a terribly good thing to do. Any opinions?

@gares, @ybertot : maybe you can comment on in how far ssreflect and the standard library should be usable in one development without issues like this. This change definitely makes this a bit harder.

Zimmi48 commented:
Cc @herbelin: he's the scope expert and probably the author of the offending change.

herbelin commented:
> Cc @herbelin: he's the scope expert and probably the author of the offending change.

As far as I know, my changes to scopes are not part of v8.11. Could it be related to changes in the semantics of `Import`? Various changes can indirectly have an impact too: for instance, the move from `fourier` to `lra` in 8.9 impacted the interpretation of numerals by removing some imports. In any case, may some additional compensating `Import` fix the issue?

Also, is there an easy way to reproduce w/o windows?

Zimmi48 commented:
Indeed, @herbelin you are probably correct with your diagnosis that this is about the changes to `Import`. @MSoegtropIMC See https://coq.github.io/doc/v8.11/refman/changes.html#exportbug and PR #10476. Almost all projects in CI had to be fixed following this change. Interval and Gappa are part of the Windows installer but not of the main CI, and that's why the problem appears only now (I don't think there is anything Windows-specific). @maximedenes may help, but he can't be the one fixing all projects: we must have some other users port projects on their own to assess the difficulty of doing so, and if the provided explanations are sufficient.

Zimmi48 commented:
And just to clarify, there was no change to the `N` delimiter: it has always been the case that it represents several distinct scopes.

MSoegtropIMC commented:
> And just to clarify, there was no change to the `N` delimiter: it has always been the case that it represents several distinct scopes.

The reason I said that is that in 8.10.2 at the same point I get for Print Scopes:

```
Scope nat_scope
Delimiting key is N
Bound to classes nat Nat.t
"x 'mod' y" := Nat.modulo x y
"m < n" := leq (S m) n

Scope N_scope
Delimiting key is num
Bound to classes N N.t
"x 'mod' y" := N.modulo x y
"x < y" := N.lt x y
```

MSoegtropIMC commented:
I guess I found what is going on:
```
/mathcomp/ssreflect/ssrnat.v:128:Delimit Scope N_scope with num
```
Probabaly this scope delimter declaration is missing cause of the Import change, so I just have to add it the Import.

herbelin commented:
Then you probably need to add an `Import ssrnat.` somewhere after other imports (since importing `ssrnat` is what overrides to `N` and `num` the default delimiters `nat` and `N` for respectively `nat_scope` and `N_scope`). One possible scenario is that you have in 8.11 a chain of imports of the form `ssrnat` (which moves the two keys), then `BinNums` (which restores `N` for `N_scope`), while in 8.10.2 there was an additional  re-import of `ssrnat` after `BinNums`.

MSoegtropIMC commented:
> Then you probably need to add an `Import ssrnat.`

Indeed, adding
```
From mathcomp.ssreflect Require Import ssrnat
```
at the end of teh imports does the trick. Patch is on the way.

herbelin commented:
> Patch is on the way.

Great!

Regarding the other question...

> @gares, @ybertot : maybe you can comment on in how far ssreflect and the standard library should be usable in one development without issues like this. This change definitely makes this a bit harder.  
                                                                              
I don't know if @gares or @ybertot intend to reply. My own perception of what ssreflect developers wish wrt this kind of clash between Coq standard library and ssreflect is that they regret them, but, at the current time, they consider important that the specificities of their design choices are respected as they are, whatever they are. See #11108 for an attempt to accomodate their wish on this particular point.

gares commented:
> how far ssreflect and the standard library should be usable in one development without issues like this

Given the root cause of the issue, I believe the question is not very pertinent anymore.
In many cases the breakages were within a single development. The bug was very unfortunate.

MSoegtropIMC commented:
A question: is there a way to remove a notation? I have various issues with notations like 
```
Notation Coq.Arith.Plus.plus_comm
Constant Coquelicot.Hierarchy.plus_comm (shorter name to refer to it in current context is Hierarchy.plus_comm)

Notation Coq.Lists.List.tail
Constant Coq.micromega.Env.tail (shorter name to refer to it in current context is Env.tail)
Constant Interval.Tactic.IntervalTactic.Private.Int'.TM.Pol.tail
  (shorter name to refer to it in current context is Int'.TM.Pol.tail)
Constant Interval.Tactic.IntervalTactic.Private.A.TaylorValuator.TM.Pol.tail
  (shorter name to refer to it in current context is A.TaylorValuator.TM.Pol.tail)
Constant Interval.Poly.Datatypes.PolR.tail (shorter name to refer to it in current context is Datatypes.PolR.tail)
```
The only way I see how this can be fixed is to use a partially or fully qualfied name, but in some cases this is tricky. Also it is not terribly elegant.

MSoegtropIMC commented:
I added a patch for Interval. Not in all places the nicest solution but back portable and it works. I couldn't work on the gappa plugin because gitlab.inria.fr is down so the version querries don't work and the scripts don't run. If the server is up tomorrow morning, the gappa fix shouldn't take long.

MSoegtropIMC commented:
Note: the Windows builds both failed cause of a server issue (could not download https://github.com/alainfrisch/flexdll/releases/download/0.37//flexdll-bin-0.37.zip). I will not restart it now because I expect to finish the Gappa plugin patch this morning and then it anyway runs again.

MSoegtropIMC commented:
@ppedrot : I still have the issue that the request
```
git ls-remote https://gitlab.inria.fr/flocq/flocq refs/heads/flocq-3.2.0
````
times out. The scripts won't run without substantial patching if a major source package server is down. Can you please inform INRIA IT to get this fixed asap (if not done so already)?

ppedrot commented:
@MSoegtropIMC it works for me, are you sure there is no firewall problem or such on your side?

ybertot commented:
It works for me too.  @ppedrot can you investigate whether the Inria IT have not a special access rule to gitlab.inria.fr for machines from Inria (it happened for some other servers)?
Anyway, I did experience a lot of connexion problems from outside Inria yesterday night (but I am not outside anymore).

MSoegtropIMC commented:
The strange thing is that it did work for me yesterday morning, but it did stop working sometime in the afternoon.

ppedrot commented:
@ybertot I am not on an Inria network right now, so it's probably not related.

ybertot commented:
Great, this settles my question then.

ppedrot commented:
@MSoegtropIMC if you have a patch likely to work, can you push it to see whether the CI passes? Thanks.

MSoegtropIMC commented:
Hmm it also doesn't work if I disable my firewall and it also doesn't work to access the web page in a browser. Let me try it from a VPN network ...

MSoegtropIMC commented:
> if you have a patch likely to work, can you push it to see whether the CI passes? Thanks.

Some things are a bit hard to figure out without a reference and right now I can't build 8.10.2 locally cause of this issue. But I guess I can fix the remainign bugs with some thinking as well ... let me see.

ppedrot commented:
FWIW, there is no branch called that way, but there is a tag indeed:
```
a9be51660cd865d323b0b3bbb9faa793856ea0de        refs/tags/flocq-3.2.0
```
You should hardwire it for now, probably.

MSoegtropIMC commented:
> FWIW, there is no branch called that way

Afair the script tries both a tag name and a branch name (it doesn't know which it is) and then draws conclusions.

Regarding access to https://gitlab.inria.fr:

What does not work:
- Access from my private network using Windows, Firefox, Gdata
- Access from my private network using Windows, git
- Access from my private network using OSX, Safari
- Access from my private network using OSX, git
- Access from my companies network using OSX, Safari
- Access from my companies network using OSX, git
- Access via varios proxy satelites in US and Germany from a paid proxy service on Windows, Firefox.

The first option did work until around yesterday noon and the next two options did work at least a week back.

What does work
- Access from my mobile phone, Android, web browser, Carrier network

If someone has a good explanation for these mysteries, please let me know ...

I try to fix gappa without refrence build.

ybertot commented:
I just checked the news from the IT department and they are aware of a problem, which they only attribute to difficult to access from the orange telecommunication and internet provider company.  I will try to make them aware that it is not only limited to orange.

MSoegtropIMC commented:
OK, Gappa plugin was trivial - only one issue and one I have seen before in Interval.

Zimmi48 commented on dev/build/windows/patches_coq/interval.patch:
```diff
@@ -0,0 +1,187 @@
+ Variable u : U.
+ 
+-Definition add := map2 (C.add u) id.
++Locate id.
```
That's probably something to remove before submitting the patch upstream.

MSoegtropIMC replied:
Indeed. There are other places to clean up. E.g. I think that the issues with ssrnat is best fixed by changing Interval's "Missing" file for Mathcomponents. Also I am looking for a way to undefined Notations which overlap ssreflect definitions - currently I give explicit names. Do you happen to know a way how to undefined notations?

I hope for a beta it is OK as is.

herbelin replied:
Hi, there is actually already code for undefining notations but no syntax for it. I'll make a proposal hopefully today or tomorrow.

MSoegtropIMC replied:
> Hi, there is actually already code for undefining notations but no syntax for it. I'll make a proposal hopefully today or tomorrow.

It would be great if we could get this into 8.11 (non beta). I think many of the issues in Interval could then be solved by undefiining certain new notations, possibly in a SSreflect compatibility file, which possibly should be part of the standard library (undefined new notations 8.X). The import issue I think can be fixed at a central place in interval, so that only very few lines would have to be changed.

Btw.: The already reported bug that Print Visibility does an anomaly if SSreflect is loaded doesn't make debugging such issues easier - it would be good to fix this in 8.11.

herbelin replied:
There is a fix for the `Print Visibility` bug at #11276.

I have a branch for deactivating notations but it is not fully clear what user API we want for it.

For instance, assuming we have commands of the form:
```
Deactivate "_ + _" : nat_scope.
Reactivate "_ + _" : nat_scope.
Deactivate pred_of_simpl.
Reactivate pred_of_simpl.
```
Do we want them to have an effect at require-time of the .vo file in which they occur? Or do we want to consider them only as "interactive" features with an effect only over parsing/printing in the current document.

MSoegtropIMC replied:
The main purpose for disabling notations would be having a way to fix compatibility issues by hiding names in e.g. ssreflect with new notations from the standard library. This would only work in a comfortable way if the deactivation of standard library notations could be done by importing ssreflect.

MSoegtropIMC commented:
Sorry, a silly error - I moved the gappa plugin patch to the wrong folder. It should work now.

ppedrot commented:
Looks like it worked, this is ready to merge right?

MSoegtropIMC commented:
> Looks like it worked, this is ready to merge right?

From my point of view yes. The minor issues in the patches are irrelevant because this will be fixed upstream anyway. The patches are just temporary for the beta. But I can't approve this since I did part of it. Unfortunately I don't have time to test the installer before Thursday evening, but it is unlikely that there are severe issues.

ppedrot commented:
There doesn't seem to be a huge enthusiasm to test the Windows binaries on Gitter, hence I think I'll wait for a summary test from @MSoegtropIMC before tagging. Better safe than sorry...

MSoegtropIMC commented:
OK, I will test the CI artefacts from this PR today evening and report here.

MSoegtropIMC commented:
There are a few issues:

- certain calls to the intervale tactic (likely many) fail cause of the Import issue - some name hiding issues are still there because they appear in tactics and are not seen until the tactic is executed - not nice. This is a regression vs. 8.10.2.

- In CoqIDE the compile buffer function does not work. The argument of topfile may not contain a colon:
```
coqc -topfile D:\test\somefile.v -Q D:\test  "D:\test.v"
Error:
Invalid character ':' in identifier "D:\test\somefile".
```
This is not a regression - it is the same in 8.10.2

- I have a file where CoqIDE crashes immediately reproducibly on open. This is also not a regression - it is the same in 8.10.2. Unfortunately I couldn't test the gappa plugin, because this applies to all simple files using the gappa plugin - probably not a coincidence.

I guess we need better automated testing for such things.

The Interval tactic issues I can fix until Monday, but it is hard to tell if I get all of them - since Interval has quite complex Ltac scripts this probably can only be guaranteed by deep testing. The CoqIDE gappa crash issue I need to look into - I can't tell what this is.

ppedrot commented:
Thanks. The beta1 have been tagged, if the regression you mention is deemed too important we may release another beta.
